From fdce58e4898510db6b6913bd622ab06621f1e9ee Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andreas=20M=C3=BCller?= <schnitzeltony@gmail.com>
Date: Sat, 4 Jan 2020 01:56:25 +0100
Subject: [PATCH] Do not check fo zenity executable
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Upstream-Status: Inappropriate [OE-Specific]

Signed-off-by: Andreas Müller <schnitzeltony@gmail.com>
---
 configure.ac | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/configure.ac b/configure.ac
index 919a7d7..780d2d4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -450,11 +450,6 @@ fi
 
 AC_SUBST(GDK_PIXBUF_CSOURCE)
 
-AC_PATH_PROG(ZENITY, zenity, no)
-if test x"$ZENITY" = xno; then
-  AC_MSG_ERROR([zenity not found in your path - needed for dialogs])
-fi
-
 AC_ARG_ENABLE(debug,
 	[  --enable-debug		enable debugging],,
 	enable_debug=no)
-- 
2.21.0

